<?xml version="1.0"?>
<launch>

  <!-- Arguments -->
  <arg name="enable_rviz" default="true"/>
  <arg name="rviz_config" default="$(find dwa)/rviz/dwa.rviz"/>

  <!-- ===== PointCloud2 to LaserScan 변환 ===== -->
  <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
    <!-- 입력 PointCloud2 토픽 -->
    <remap from="cloud_in" to="/ouster/points"/>
    <remap from="scan" to="/scan"/>

    <!-- 변환 파라미터 -->
    <param name="target_frame" value="os_lidar"/>
    <param name="transform_tolerance" value="0.01"/>
    <param name="min_height" value="-0.5"/>
    <param name="max_height" value="2.0"/>

    <param name="angle_min" value="-3.14159"/>
    <param name="angle_max" value="3.14159"/>
    <param name="angle_increment" value="0.0087"/>
    <param name="scan_time" value="0.1"/>
    <param name="range_min" value="0.3"/>
    <param name="range_max" value="75.0"/>

    <param name="use_inf" value="true"/>
    <param name="inf_epsilon" value="1.0"/>
  </node>

  <!-- ===== move_base 네비게이션 스택 ===== -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <!-- 파라미터 로드 -->
    <rosparam file="$(find dwa)/config/costmap_common_params.yaml" command="load" ns="global_costmap"/>
    <rosparam file="$(find dwa)/config/costmap_common_params.yaml" command="load" ns="local_costmap"/>
    <rosparam file="$(find dwa)/config/move_base_params.yaml" command="load"/>
    <rosparam file="$(find dwa)/config/global_costmap_params.yaml" command="load"/>
    <rosparam file="$(find dwa)/config/local_costmap_params.yaml" command="load"/>
    <rosparam file="$(find dwa)/config/dwa_local_planner_params.yaml" command="load"/>
    <rosparam file="$(find dwa)/config/global_planner_params.yaml" command="load"/>

    <!-- 플래너 선택 -->
    <param name="base_global_planner" value="global_planner/GlobalPlanner"/>
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>

    <!-- cmd_vel 토픽 리매핑 (Husky 컨트롤러로 전달) -->
    <remap from="cmd_vel" to="husky_velocity_controller/cmd_vel"/>
  </node>

  <!-- ===== RViz 시각화 ===== -->
  <node if="$(arg enable_rviz)" name="rviz" pkg="rviz" type="rviz"
        args="-d $(arg rviz_config)"
        output="screen"/>

</launch>
